infoschema/slow_log_cache.go

```diff
@@ -0,0 +1,337 @@
+// limitations under the License.
+
+package infoschema
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+var (
+	defSlowQueryBufferSize = 500000
+	slowQueryBufferSize    = defSlowQueryBufferSize
+)
+
+var globalSlowQueryReader slowLogReader
+
+// slowLogReader uses to read slow log data from slow log file.
+// It will cache the data in a ring buffer to avoid repeated read file and parse data.
+type slowLogReader struct {
+	sync.RWMutex
+	// TODO: support adjust cache size.
+	cache *slowLogCache
+}
+
+// getSlowLogData uses to get slow log data.
+// It will try to get from cache data first.
+// If cache not match, it will get data by parse slow log file, and then update the cache by async.
+func (s *slowLogReader) getSlowLogData(filePath string, tz *time.Location) ([][]types.Datum, error) {
+	if s.cacheValid(filePath) {
+		// try read with buffer.
+		return s.getSlowLogDataWithCache(tz)
+
+	}
+	return s.parseSlowLogFileAndUpdateCache(tz, filePath, 0, nil, nil)
+}
+
+// getSlowLogDataWithCache gets data from cache and get the remain data by parse slow log file.
+// It will update cache if has new data.
+func (s *slowLogReader) getSlowLogDataWithCache(tz *time.Location) ([][]types.Datum, error) {
+	s.RLock()
+	// Get before cached tuples.
+	beforeCacheTuples, hasTruncate, err := s.parseSlowLogFileBeforeCachedAndCheckTruncate(tz)
+	if err != nil {
+		s.RUnlock()
+		return nil, err
+	}
+	// Try to avoid reallocate memory.
+	rows := make([][]types.Datum, len(beforeCacheTuples), s.cache.buf.len()+len(beforeCacheTuples)+100)
+	// get all cache tuples.
+	rows = s.getSlowLogRowsFromCache(tz, rows)
+	cacheEndTime := s.cache.getEndTime()
+	readOffsetAfterCache := s.cache.getEndOffset()
+	s.RUnlock()
+
+	if hasTruncate {
+		readOffsetAfterCache = 0
+	}
+	// Get after cached tuples. Put this out lock is for reduce lock time.
+	afterCacheTuples, err := s.parseSlowLogFileAfterCached(tz, cacheEndTime, readOffsetAfterCache)
+	if err != nil {
+		return nil, err
+	}
+	logutil.Logger(context.Background()).Info("slow query read data with cache", zap.Int("before cached", len(beforeCacheTuples)), zap.Int("cached", len(rows)), zap.Int("after cached", len(afterCacheTuples)))
+
+	// Fill before cached tuples.
+	for i := range beforeCacheTuples {
+		rows[i] = beforeCacheTuples[i].convertToDatumRow()
+	}
+	// Append after cache tuples.
+	for i := range afterCacheTuples {
+		rows = append(rows, afterCacheTuples[i].convertToDatumRow())
+	}
+	// Update cache.
+	go s.updateCache(s.cache.filePath, afterCacheTuples)
+	return rows, nil
+}
+
+// parseSlowLogFileAndUpdateCache gets data by parse the slow log file, and then update the cache.
+func (s *slowLogReader) parseSlowLogFileAndUpdateCache(tz *time.Location, filePath string, offset int64, filterFn func(t time.Time) bool, bypassFn func(t time.Time) bool) ([][]types.Datum, error) {
```

Comment:
`filterFn` and `bypassFn` are always nil, so why don't remove them?

Reply:
great, thanks.